### DISCUSSION TITLE
Peppol Reporting missing C1 country code for incoming items

### DISCUSSION THREAD

**Discussion by Jef-VDD** [phax/phase4#208] [General]
Hello @phax 

Since 01.01.2024 when creating a reporting item for an incoming message that is missing their peppolSBD c1CountryCode we see logs a warning log:
> C1 Country Code is missing

followed by error:
> Not all mandatory fields are set. Cannot create Peppol Reporting Item

This is due to to the following statement that activated since 2024 in the **PeppolReportingItem** builder.
```      // m_sCountryC1 may be null, before 1.1.2024
      if (PDTFactory.getCurrentLocalDateUTC ().compareTo (DATE_COUNTRY_C1_BECOMES_MANDATORY) >= 0)
      {
        if (StringHelper.hasNoText (m_sCountryC1))
        {
          LOGGER.warn ("The field 'countryC1' is not set");
          return false;
        }
      }
```
      
I recall from the SPC meetings if the C2 Service provider does not provide a C1 Country code we should report it as country code 'ZZ' from the following codelist https://docs.peppol.eu/edelivery/specs/reporting/tsr/trn-transaction_statistics/codelist/ISO3166/. However I can't seem to find this in any other documentation.

In short I would just like to know what the proposed way to handle missing c1CountryCodes are in the Phase4 implementation of the access point. Is it possible for the builder to fill in 'ZZ' if the C1 code is missing Or should the C3 accesspoint alter the incoming **peppolSBD** with country code ZZ that is handed over to **Phase4PeppolServletMessageProcessorSPI.createPeppolReportingItemForReceivedMessage** ?

Thanks for you work within the peppol space and a happy new year!

**Comment by phax** [comment]
Thanks for asking the question :)

From a `PeppolReportingItem` perspective, the country code is mandatory. As such this change is totally okay imho.

However, the change needs to happen at the place where the collection takes place. Based on the email from OpenPeppol you have basically two options when receiving a messages without a country code:
1. Reject the message, based on non-compliance or
2. Accept the message, but use `ZZ` as the "assumed country code of C1"

In scenario 1 you don't need a `PeppolReportingItem`.
In scenario 2 you need to pass the `ZZ` to the `PeppolReportingItem`

Hope that makes sense

**Reply by Jef-VDD** [reply]
Hey @phax, thanks for the quick response and clarification!
    
In Scenario 2 I am using a custom implementation for the `IPhase4PeppolIncomingSBDHandlerSPI` and after my business logic i hand over the `Ebms3UserMessage, PeppolSBDHDocument, IAS4MessageState` to the `Phase4PeppolServletMessageProcessorSPI.createPeppolReportingItemForReceivedMessage` method provided by `phase4-peppol-servlet` dependency.

This will however return `NULL` when no c1CountryCode is present in the `PeppolSBDHDocument` and I see no way to provide the country code when using `createPeppolReportingItemForReceivedMessage`.
Our current solution is to implement the `PeppolReportingItem.builder` ourselves when c1CountryCode is not provided bypassing `createPeppolReportingItemForReceivedMessage`.  But I was wondering if there isn't a way to still make use of the `createPeppolReportingItemForReceivedMessage` method since I would prefer to use Phax libraries :)

Ps. I assume that when I would like to implement scenario 1 I could just throw a `Phase4PeppolClientException` in my custom `IPhase4PeppolIncomingSBDHandlerSPI`?

**Reply by phax** [reply]
In the update of phase4 2.7.0 the country C1 is mandatory by default.

The sender is basically forcing you to use COUNTRY_C1 - there are hacky ways to work around, but I won't explain them here :)

On the receiving side, it is made mandatory by default now, but can be triggered via this global switch:
```java
   // Our server expects all SBDH to contain the COUNTRY_C1 element in SBDH
    // (this is the default setting, but added it here for easy modification)
    Phase4PeppolServletConfiguration.setCheckSBDHForMandatoryCountryC1 (true);
```

For the reporting builder on the receiver side, simply use code like `countryC1 = hasText(sbdh.countryC1) ? sbdh.countryC1 : "ZZ"` - this is your fallback. The fallback is not done automatically in the reporting lib
